### ISSUE TITLE
I get an error when I import the csvm file.

### ISSUE THREAD

**Issue by haiyunaixueye** [SEI-AMS/pycloud#9]
Hello, I found that when importing the csvm file, there will be unpacking errors. This problem will directly affect the next operation of service instantiation. This error is as follows:

![image](https://user-images.githubusercontent.com/24189548/38668255-4cf3ffe4-3e76-11e8-96ff-36632f123bcc.png)

I found out by debugging that there was a problem with line 259 of the file, ~\pycloud\pycloud\model\servicevm.py
Line 99 of the file,  ~\pycloud\pycloud\vm\vmsavedstate.py,

Looking forward your reply.

**Comment by sebastian-echeverria** [comment]
That error is expected for csvm files that do not contain a memory image. However, it is not being handled properly for some reason. Let me try to get that to happen locally.

A quick question: has this always happened to you? Or are you re-importing a service that you had successfully imported before? If this is the case, you may get around the issue by first removing the old service in the Manager, and then importing it again.

**Comment by haiyunaixueye** [comment]
This has always been the case. I have imported successfully, but only once. And then i never succeed.
I have found the file, .lqs, is 0 bytes.

Looking forward your reply.

**Comment by sebastian-echeverria** [comment]
I see the bug btw, and it should be easy to fix, but I need to set up the test environment first to ensure it works. Two questions: have you tried removing the service and then importing it again? And how did you create the csvm file you are using?

**Comment by haiyunaixueye** [comment]
I repeat it many times and still can't.
I use the csvm file that you provide.

**Comment by sebastian-echeverria** [comment]
Yes, but my question was whether you removed the service first, with the X button on the Available Services page, and only then tried to import again. But I will assume that is what you tried.

Can you remind me what is the CSVM file I provided you? Was it in a link? If you change the extension of that csvm file to .tar.gz, and decompress it, what are the files inside it?

Again, I can see the bug, but I am not able to reproduce it.... and if I can't reproduce it, I won't really be sure if I fix it when I make the changes. I just want to get as much information as I can about your current inputs so as to reproduce the error on my environment. I will try to manually create a CSVM file that has a lqs file inside it of size 0, but I am not sure how that can be created by any tool right now...

**Comment by sebastian-echeverria** [comment]
Ok, I was able to manually create a .csvm file with a zero-length lqs file, and with that I was able to reproduce the error. The issue should be fixed now in the master branch. Do try it out and let me know if it works now.

However, I am still unsure on how you were able to create or obtain a CSVM file with a zero-length lqs file. I am wondering if there is a bug somewhere else that could lead to this. In any case, this type of csvm file should be properly handled now.

**Comment by haiyunaixueye** [comment]
Thank you very much. The csvm file can now be imported normally.
I used the speech-server code(https://github.com/SEI-AMS/speech-server)  to try to generate the csvm file, but failed, so I used the csvm file you provided. I unzipped the csvm file you provided. I found that the state_image in the service.json file does not match the name of the .lqs file,so I modified the service.json file and then repackaged it to generate the csvm file. Importing is still wrong(pycloud 3.0..0)

1. I added csvm file, first delete the original available service, then import csvm
2. The csvm file I used is the one you provided in bug #6. As shown below：
![image](https://user-images.githubusercontent.com/24189548/38712822-aab0733a-3f00-11e8-9104-5b80ce8e289b.png)
3. I unzipped the csvm file and got three files, as shown in the figure below:
![image](https://user-images.githubusercontent.com/24189548/38712838-cccd1f4a-3f00-11e8-8838-6e7863a54dfe.png)

**Comment by sebastian-echeverria** [comment]
I see, thanks for reminding me. I may have partially created that file manually, I remember I did not have full access to the dev environment at that point. I recreated the csvm file now with the current scripts and now there is no lqs file in there. But in any case, after this fix, it doesn't really matter. Thanks for helping us fix the issue! In any case the lqs file used to be a necessary file inside the csvm but it is no longer needed when importing; thus the issues we saw.

By the way, if you want to I can try again to help you create an csvm file with the capstan utility. I have full access to a dev environment right now so we can try to debug the issues you had before. This would simplify your testing process, in case you want to modify the models for speech-server and repackage and re-import the csvm files quickly into the cloudlet.

**Comment by haiyunaixueye** [comment]
Thank you, we used to generate a csvm file based on the speech server (https://github.com/SEI-AMS/speech-server)，but failed，and maybe there were  some  problems with my steps.If you can give me  a tutorial to generate a csvm file ,I will be very grateful.

**Comment by sebastian-echeverria** [comment]
Ok, I updated some of the scripts that create the file for speech in the speech-server repo, since some of them were a bit outdated and were not installing the best version to use. I also wrote a more detailed explanation on how to create the CSVM file here: https://github.com/SEI-AMS/speech-server/wiki/Creating-a-Speech-CSVM . In practice, you should only need to run two scripts, but there are more details in that page about what is going on underneath.

**Comment by sebastian-echeverria** [comment]
I will close this for now. If you have further questions regarding CSVM creation, please do create a new issue and we can discuss it there.

**Comment by haiyunaixueye** [comment]
I'm sorry that I forget to  reply you  in time ,In fact, running the script can generate a csvm file and I can add it successfully. However, this service is not available. That is, when connecting to Android, it cannot process voice files, as shown in the figure below.
![image](https://user-images.githubusercontent.com/24189548/38840873-435d4e30-4214-11e8-91c7-d8db2821bb6e.png)

**Comment by haiyunaixueye** [comment]
Also want to ask you, what are the development environments for these projects?

**Comment by sebastian-echeverria** [comment]
Ok. So you say that with the previous CSVM file from my link, everything worked. But when you create a new one with Capstan, the Speech Server does not reply. Right? I will look into it and let you know what I find out.

We use a laptop with Ubuntu 16.04 (or 16.10, not sure right now), plus a wierless AP, which is connected to the laptop through an Ethernet cable. If you want to know anything else that is more specific, just let me know.

**Comment by haiyunaixueye** [comment]
Yes，when Added a newly created csvm file,something was wrong with the data returned by the server. I know the speech client uses Android Studio development. Is the speech server also?

**Comment by sebastian-echeverria** [comment]
We have used Android Studio for both Speech Client and Cloudlet Client, but it is not mandatory. Both are Gradle projects. Speech Server us just a Java Gradle project. I have used IntelliJ Idea to play with it, but you can use any IDE you want. Or just the command line with Gradle.

**Comment by haiyunaixueye** [comment]
Ok,Thank you very much ,and my next step is to take a close study for the code.

**Comment by sebastian-echeverria** [comment]
Ok, I looked into the CSVM file that was being generated with capstan, and there were two issues with it. One was that the Sphinx library had changed a bit since we last built this, so some configurations and classes were slightly different. I updated Speech Server to version 5.1.1 (now in master branch), so now this should work properly. Btw, you may have seen these issues too if you just tried to send files to Speech Server directly.

The other error was on the VM creation on Pycloud. For some reason, the current CPU configuration when creating a new VM was making the VM sometimes get stuck when loading, if the os was OSv (which is the case for the Speech csvm file we are using). I fixed this and it should now be fixed in master (v 3.0.2).

So, what you should do to test this out:

- Pull the most recent version of master for speech-server
- Re-create the .csvm file
- Pull the most recent version of master for pycloud
- Start Cloudlet Manager
- Delete the existing service
- Re-import the new .csvm speech file

Things should work properly now. Do let me know if there are still issues.

**Comment by haiyunaixueye** [comment]
Thanks for your answer. The lastest version of csvm is working fine. However, there is still a minor problem, that is, the service just started and the speech client can't receive data from the server. I need to upload audio files several times , otherwise my Android client won't receive data.

I found a new bug, the pip program has now been upgraded to 10.0, so the module will not be found during setup.py.  The changlog of pip10 is shown below:

![image](https://user-images.githubusercontent.com/24189548/39079496-a2ab3ad6-454e-11e8-830c-6d029228de0e.png)

**Comment by sebastian-echeverria** [comment]
So the error now is that it takes a while to start working, as in you have to wait some seconds before requests will actually get a reply? Or you have to try several times before you get a reply? Could you define exact steps to reproduce this? I have not seen this behavior in my quick tests. How long do you have to wait before getting a real reply? Or how many times do you have to try before getting a reply?

I will take a look into the pip issue. Thanks!

**Comment by haiyunaixueye** [comment]
Yes.Every time before the speech client gets the correct result, it must try to click “send to server” multiple times, and that  is  subject to change. In short, in all my test, clicking  once “send to server”button has never been successful . The time when the result was received is not  very long.

**Comment by sebastian-echeverria** [comment]
Have you tried waiting, say, 10 or 15 seconds after connecting to the pycloud service, before clicking on the Send to Server button? I am trying to figure out if it is an issue related to the time it takes the VM to be up and running.

**Comment by haiyunaixueye** [comment]
Yes，I had been waiting until I return  "An IO Exeption has occured..."information in the speech client
The results of the most recent test are shown in the following picture.
![image](https://user-images.githubusercontent.com/24189548/39135790-59cf979e-474c-11e8-8a9d-5b97603bfe64.png)

**Comment by sebastian-echeverria** [comment]
Thanks! Regarding the issue about sending several times before it works, I still have to look at it in more detail, but my guess is that the VM is sometimes taking to long when starting up. I will let you know if I can get to the issue and fix it.

Regarding the pip issue, it is correct that by moving all of their APIs to another package they broke existing setup.py code, so there are issues with pip 10. I updated the setup scripts to fix that, and also to fix some minor incompatibilities with the use of systemd in Ubuntu 16 (which is the current min OS I am testing on), in case you want to check that out.

**Comment by sebastian-echeverria** [comment]
Closing since all bugs were fixed, except the one that was moved over to issue #12.
